js-packages/@prestojs/viewmodel/src/ViewModelCache.ts

```diff
@@ -9,10 +9,49 @@ import ViewModel, { PrimaryKey } from './ViewModel';
  * to just check if it's an instanceof RecordPointer
  */
 class RecordPointer<T extends ViewModel> {
+    /**
+     * This is the value that was replaced by this pointer in the cache. When we actually clone
+     * the record we compare against this value - if it's the same we return the original object
+     * so that equality checks still hold. This is because we never return the RecordPointer - we
+     * only resolve the underlying value as is required. In order to determine whether or not
+     * an object is the same requires first cloning it and then comparing it. As such we do that
+     * at the time we need the cloned value, not upfront (ie. we don't do it before we store the
+     * RecordPointer as that would be lower than doing it lazily - especially if you are updating
+     * lots of records)
+     */
+    currentValue?: null | T;
     record: T;
-    constructor(record: T) {
+    constructor(currentValue: null | undefined | T | RecordPointer<T>, record: T) {
+        this.currentValue =
+            currentValue instanceof RecordPointer ? currentValue.currentValue : currentValue;
         this.record = record;
     }
+
+    clone(fieldNames: string[]): T {
+        const cloned = this.record.clone(fieldNames);
```

Comment:
add comment indicating we clone first as field names may be subset of original record field names - we need to compare only this subset below